experimental/lib/Transforms/LSQSizing/LSQSizingSupport.cpp

```diff
@@ -0,0 +1,574 @@
+
+  for (Operation &op : funcOp.getOps()) {
+    // Get operation's basic block
+    unsigned srcBB;
+    if (auto optBB = getLogicBB(&op); !optBB.has_value())
+      continue;
+    else
+      srcBB = *optBB;
+
+    // The basic block the operation belongs to must be selected
+    if (!cfdfcBBs.contains(srcBB))
+      continue;
+
+    // Add the unit and valid outgoing channels to the CFDFC
+    addNode(&op, extractNodeLatency(&op, timingDB));
+
+    for (OpResult res : op.getResults()) {
+      assert(std::distance(res.getUsers().begin(), res.getUsers().end()) == 1 &&
+             "value must have unique user");
+
+      // Get the value's unique user and its basic block
+      Operation *user = *res.getUsers().begin();
+      unsigned dstBB;
+      if (std::optional<unsigned> optBB = getLogicBB(user); !optBB.has_value())
+        continue;
+      else
+        dstBB = *optBB;
+
+      if (srcBB != dstBB) {
+        // The channel is in the CFDFC if it belongs belong to a selected arch
+        // between two basic blocks
+        for (size_t i = 0; i < cfdfcBBs.size(); ++i) {
+          unsigned nextBB = i == cfdfcBBs.size() - 1 ? 0 : i + 1;
+          if (srcBB == cfdfcBBs[i] && dstBB == cfdfcBBs[nextBB]) {
+            addChannelEdges(res);
+            if (buffer::CFDFC::isCFDFCBackedge(res))
+              addChannelBackedges(res, (II * -1));
+            break;
+          }
+        }
+      } else if (cfdfcBBs.size() == 1) {
+        // The channel is in the CFDFC if its producer/consumer belong to the
+        // same basic block and the CFDFC is just a block looping to itself
+        addChannelEdges(res);
+        if (buffer::CFDFC::isCFDFCBackedge(res))
+          addChannelBackedges(res, (II * -1));
+      } else if (!isBackedge(res)) {
+        // The channel is in the CFDFC if its producer/consumer belong to the
+        // same basic block and the channel is not a backedge
+        addChannelEdges(res);
+      }
+    }
+  }
+}
+
+void AdjListGraph::addNode(mlir::Operation *op, int latency) {
+  nodes.insert(
+      {getUniqueName(op).str(), AdjListNode{latency, -1, op, {}, {}, {}}});
+}
+
+void AdjListGraph::addEdge(mlir::Operation *src, mlir::Operation *dest) {
+  nodes.at(getUniqueName(src).str())
+      .edges.insert(
+          getUniqueName(dest).str()); // Add edge from node u to node v
+}
+
+void AdjListGraph::addChannelEdges(mlir::Value res) {
+  mlir::Operation *srcOp = res.getDefiningOp();
+  for (Operation *destOp : res.getUsers()) {
+    addEdge(srcOp, destOp);
+  }
+}
+
+void AdjListGraph::addChannelBackedges(mlir::Value res, int latency) {
+  mlir::Operation *srcOp = res.getDefiningOp();
+  for (Operation *destOp : res.getUsers()) {
+    addBackedge(srcOp, destOp, latency);
+  }
+}
+
+void AdjListGraph::printGraph() {
+  for (const auto &pair : nodes) {
+    std::string opName = pair.first;
+    const AdjListNode &node = pair.second;
+    llvm::dbgs() << opName << " (lat: " << node.latency
```

Comment:
Unless you have a good reason, you should use `llvm::errs()` for this kind of print statements. AFAIK `llvm::dbgs()` is more supposed to work in tandem with `LLVM_DEBUG` macros and such.